File: src/tests/comparison.rs
use crate::wave::{Phase, WavePacket};
use crate::field::ResonanceField;

// =============================================================================
// Test 8: Wave vs Linear Scan (Correctness)
// =============================================================================
pub fn test_8_wave_vs_linear() -> bool {
    println!("--- Test 8: Wave vs Linear Scan (Correctness) ---");

    let bucket_count = 100;
    let entity_count = 1000;
    let target_value: u64 = 42;

    // Generate entities with deterministic "random" values
    let mut field = ResonanceField::new();
    let mut values: Vec<(String, u64)> = vec![];

    for i in 0..entity_count {
        // Simple deterministic hash to spread values
        let val = (i * 37 + 13) % bucket_count as u64;
        let id = format!("entity_{i}");
        values.push((id.clone(), val));
        let packet = WavePacket::new(&id)
            .with_attr("value", Phase::from_value(val, bucket_count));
        field.add(packet);
    }

    // Method A: Linear scan with value comparison
    let linear_results: Vec<String> = values.iter()
        .filter(|(_, v)| *v == target_value)
        .map(|(id, _)| id.clone())
        .collect();

    // Method B: Coherence scan
    // With 100 buckets, each is 3.6°. cos(3.6°)=0.998, so threshold must be > 0.998
    // to avoid catching adjacent buckets. Use 0.9999 for single-bucket precision.
    let target_phase = Phase::from_value(target_value, bucket_count);
    let wave_results: Vec<String> = field.query_exact("value", &target_phase, 0.9999)
        .iter()
        .map(|(p, _)| p.id.clone())
        .collect();

    println!("  {} entities, bucket_count={}, target_value={}", entity_count, bucket_count, target_value);
    println!("  Linear scan found: {} matches", linear_results.len());
    println!("  Wave scan found:   {} matches", wave_results.len());

    // Check both sets are identical
    let mut linear_sorted = linear_results.clone();
    let mut wave_sorted = wave_results.clone();
    linear_sorted.sort();
    wave_sorted.sort();

    let pass = linear_sorted == wave_sorted;

    if !pass {
        println!("  MISMATCH!");
        println!("  Linear: {:?}", &linear_sorted[..linear_sorted.len().min(10)]);
        println!("  Wave:   {:?}", &wave_sorted[..wave_sorted.len().min(10)]);
    } else {
        println!("  Result sets are IDENTICAL");
    }

    println!("  RESULT: {}\n", if pass { "PASS" } else { "FAIL" });
    pass
}

// =============================================================================
// Test 9: Harmonic Query vs JOIN (Value Proposition)
// =============================================================================
pub fn test_9_harmonic_vs_join() -> bool {
    println!("--- Test 9: Harmonic Query vs JOIN ---");

    let mut field = ResonanceField::new();

    // 4 groups of 25 entities, at 0°, 90°, 120°, 240° (with slight scatter)
    let group_centers = [0.0, 90.0, 120.0, 240.0];
    let group_names = ["group_0", "group_90", "group_120", "group_240"];

    for (gi, &center) in group_centers.iter().enumerate() {
        for j in 0..25 {
            // Small scatter within ±2.4° of center
            let offset = (j as f64 - 12.0) * 0.2;
            let deg = center + offset;
            let id = format!("{}_{}", group_names[gi], j);
            let packet = WavePacket::new(&id)
                .with_attr("pos", Phase::from_degrees(deg));
            field.add(packet);
        }
    }

    // Target entity: at 3° (in the 0° group)
    let target = Phase::from_degrees(3.0);

    // Harmonic scan: 3rd harmonic finds both 120° and 240° automatically
    let harmonic_results = field.query_harmonic("pos", &target, 3, 0.85);

    // Categorize results by group
    let mut group_counts = [0usize; 4];
    for (p, _) in &harmonic_results {
        for (gi, name) in group_names.iter().enumerate() {
            if p.id.starts_with(name) {
                group_counts[gi] += 1;
            }
        }
    }

    println!("  Target: entity at 3° (in group_0)");
    println!("  3rd harmonic scan results by group:");
    for (gi, name) in group_names.iter().enumerate() {
        println!("    {}: {} entities found", name, group_counts[gi]);
    }
    println!("  Total harmonic matches: {}", harmonic_results.len());

    let found_120 = group_counts[2] > 0;
    let found_240 = group_counts[3] > 0;
    let not_found_90 = group_counts[1] == 0;

    println!("  Found group_120: {}", found_120);
    println!("  Found group_240: {}", found_240);
    println!("  Excluded group_90: {}", not_found_90);
    println!("  Key insight: Single harmonic scan found BOTH 120° and 240° groups");
    println!("  Traditional approach would need 2 JOINs for the same result");

    let pass = found_120 && found_240 && not_found_90;
    println!("  RESULT: {}\n", if pass { "PASS" } else { "FAIL" });
    pass
}

File: python/tests/comparison.py
"""Tests 8-9: Wave vs linear scan, harmonic vs JOIN."""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from wave import Phase, WavePacket
from field import ResonanceField


def test_8_wave_vs_linear() -> bool:
    print("--- Test 8: Wave vs Linear Scan (Correctness) ---")

    bucket_count = 100
    entity_count = 1000
    target_value = 42

    field = ResonanceField()
    values = []

    for i in range(entity_count):
        val = (i * 37 + 13) % bucket_count
        id_ = f"entity_{i}"
        values.append((id_, val))
        packet = WavePacket(id_).with_attr("value", Phase.from_value(val, bucket_count))
        field.add(packet)

    # Method A: Linear scan with value comparison
    linear_results = sorted([id_ for id_, v in values if v == target_value])

    # Method B: Coherence scan
    target_phase = Phase.from_value(target_value, bucket_count)
    wave_results = sorted([p.id for p, _ in field.query_exact("value", target_phase, 0.9999)])

    print(f"  {entity_count} entities, bucket_count={bucket_count}, target_value={target_value}")
    print(f"  Linear scan found: {len(linear_results)} matches")
    print(f"  Wave scan found:   {len(wave_results)} matches")

    passed = linear_results == wave_results

    if not passed:
        print("  MISMATCH!")
        print(f"  Linear: {linear_results[:10]}")
        print(f"  Wave:   {wave_results[:10]}")
    else:
        print("  Result sets are IDENTICAL")

    print(f"  RESULT: {'PASS' if passed else 'FAIL'}\n")
    return passed


def test_9_harmonic_vs_join() -> bool:
    print("--- Test 9: Harmonic Query vs JOIN ---")

    field = ResonanceField()

    group_centers = [0.0, 90.0, 120.0, 240.0]
    group_names = ["group_0", "group_90", "group_120", "group_240"]

    for gi, center in enumerate(group_centers):
        for j in range(25):
            offset = (j - 12.0) * 0.2
            deg = center + offset
            id_ = f"{group_names[gi]}_{j}"
            packet = WavePacket(id_).with_attr("pos", Phase.from_degrees(deg))
            field.add(packet)

    target = Phase.from_degrees(3.0)
    harmonic_results = field.query_harmonic("pos", target, 3, 0.85)

    group_counts = [0, 0, 0, 0]
    for p, _ in harmonic_results:
        for gi, name in enumerate(group_names):
            if p.id.startswith(name):
                group_counts[gi] += 1

    print("  Target: entity at 3 deg (in group_0)")
    print("  3rd harmonic scan results by group:")
    for gi, name in enumerate(group_names):
        print(f"    {name}: {group_counts[gi]} entities found")
    print(f"  Total harmonic matches: {len(harmonic_results)}")

    found_120 = group_counts[2] > 0
    found_240 = group_counts[3] > 0
    not_found_90 = group_counts[1] == 0

    print(f"  Found group_120: {found_120}")
    print(f"  Found group_240: {found_240}")
    print(f"  Excluded group_90: {not_found_90}")
    print("  Key insight: Single harmonic scan found BOTH 120 deg and 240 deg groups")
    print("  Traditional approach would need 2 JOINs for the same result")

    passed = found_120 and found_240 and not_found_90
    print(f"  RESULT: {'PASS' if passed else 'FAIL'}\n")
    return passed
